torch/csrc/autograd/function_port.cpp

```diff
@@ -0,0 +1,26 @@
+#include "torch/csrc/autograd/function_port.h"
+
+#include "torch/csrc/autograd/function.h"
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+namespace torch {
+namespace autograd {
+FunctionPort::FunctionPort(
+    const std::shared_ptr<Function>& function_,
+    uint32_t port_)
+    : function(function_), port(port_) {}
+
+FunctionPort::~FunctionPort() = default;
+
+bool FunctionPort::operator==(const FunctionPort& other) const noexcept {
+  return this->function == other.function && this->port == other.port;
+}
+
+bool FunctionPort::operator!=(const FunctionPort& other) const noexcept {
+  return !(*this == other);
+}
```

Comment:
Do we really need this C++ file? Again, such trivial functions are **very** likely to get inlined, and I can't see any reason to keep them out of the header (they are not even templates, there's very little build time overhead).

Reply:
I've put it into the header. But just a note/thought: I think whether or not a function inlines is premature optimization at this point. I'm very passionate about keeping header files clean and dependency (header) free unless absolutely necessary, and forward declaring as much as possible. I've worked on a few pretty large-scale, ancient C++ projects like clang/LLVM or Scribe (FB) where I experienced first-hand what bad header sanity and gradual "header rot" leads to. Including things in headers leads to people forgetting to include those headers elsewhere, creating an implicit dependency chain between header files which can span dozens of files. In the worst case, and I've spent many, many hours debugging these issues, you have something like `X.h`  including a header, then `A.h`->`B.h`->`C.h`->`D.h`->`E.h`->`F.h`->`F.cpp`. Now two horrors can occur: (1) the header `X.h` includes, happens to include `F.h`, then you get a billion line error message about undefined classes, due to cyclic includes or (2), which is *much* more common, the header `X.h` included was `<vector>` because `struct X;` uses `vector`; now `F.cpp`also uses `<vector>` and forgets to include it, relying on the dependency on `X.h`; then the implementation of `X` changes to no longer use `<vector>` and whoops suddenly `F.cpp` doesn't compile anymore. Beyond that, you have things like increased build times. `X.h` includes `Y.h` even though it could forward declare it. Now every time I change `Y.h` I have to recompile everything that includes `X.h`. In the small, all these problems seem trivial and indeed having the function and declaration in the header removes the effort of having two files open (although arguably having a "documentation/interface specification" file and a separate "implementation" file is something I always liked), but over time you get things like T23516398 where you have teams having to invest effort into untangling years of bad header hygiene. This is 100% a fault of C++ not (yet) having a module system, but until then I value header hygiene a lot. And just to give you the big picture, I ran a tool called [include-what-you-use](http://include-what-you-use.org/), a tool which checks for forgotten includes (i.e relying on other files supplying headers), over ATen and it spits out like 4000 lines of warnings: P58972845. So this is just my rationale for putting things into implementation files. That said, I get that the style in our codebase is to prefer headers, so I'll try my best to stick to that! @zdevito

Reply:
While what you're saying makes a lot of sense, and I agree with most of it, I just think that if you have a plain simple helper struct, with a trivial constructor and 2 one-liner methods, then it's an overkill to keep an extra file, especially that this code **should** be inlined.

Fortunately we're not at 30min build times, so we can afford having a more relaxed development model (although I agree that having to recompile 90 files when `ir.h` changes is a bit of a pain).